Sprint 6 solution

IvanGriko opened the pull request:
Сделать историю посещений неограниченной по размеру.
Избавиться от повторных просмотров в истории.

IvanGriko committed:
ФЗ 6.
Сделать историю посещений неограниченной по размеру.
Избавиться от повторных просмотров в истории.

IvanGriko committed:
ФЗ 6.
Сделать историю посещений неограниченной по размеру.
Избавиться от повторных просмотров в истории.

Yang-Pi commented on src/ru/yandex/tasktracker/service/Node.java:
```diff
@@ -0,0 +1,13 @@
+package ru.yandex.tasktracker.service;
+
+public class Node<Task> {
```
Если делаешь класс дженериком, то стоит указать в качестве параметра неопределенный тип T, чтобы ноду можно было использовать для любого класса. 
А вот в месте объявления ноды (head и tail) уже указывать конкретный тип параметра - Task

Yang-Pi commented on src/ru/yandex/tasktracker/service/InMemoryHistoryManager.java:
```diff
@@ -3,27 +3,93 @@
+    @Override
+    public void remove(Integer id) {
+        if (historyMap.containsKey(id)) {
+            getTasks().remove(id);
```
нет смысла удалять таску из списка, который возвращается методом getTasks(). Как ты понимаешь, этот метод возвращает каждый раз новую копию листа задач из связанного списка. Поэтому после того как ты удалишь ноду методом removeNode(), метод getTasks() автоматически будет возвращать обновленный лист, так как он собирает таски по нодам.

Yang-Pi commented on src/ru/yandex/tasktracker/service/InMemoryHistoryManager.java:
```diff
@@ -3,27 +3,93 @@
+
+    @Override
+    public void remove(Integer id) {
+        if (historyMap.containsKey(id)) {
```
метод Map#remove возвращает удаленное значение, либо null если его не было в мапе. Поэтому чтобы лишний раз не проверять наличие ключа в мапе, можно его сразу удалить методом remove и если результат последнего метода не null, то вызвать метод removeNode с этим результатом

Yang-Pi commented on src/ru/yandex/tasktracker/Main.java:
```diff
@@ -6,91 +6,68 @@
+import java.util.ArrayList;
+import java.util.List;
+
 public class Main {
```
не хватает файлов .github/workflows из задания - нужны проверки автотестами 
![image](https://github.com/user-attachments/assets/66b7c80f-2832-4b61-8e27-ff34f9590fb3)

Yang-Pi commented on src/ru/yandex/tasktracker/service/InMemoryHistoryManager.java:
```diff
@@ -3,27 +3,93 @@
+    private void removeNode(Node node) {
+        Node<Task> prevNode = node.prev;
+        Node<Task> nextNode = node.next;
+        if (size == 1) {
```
лучше не пользоваться переменной size, так как её значение может быть случайно нарушено в программе - из-за этого алгоритм становится более хрупким. Чтобы избежать лишних проверок количества элементов в списке, можно воспользоваться техникой "пустая голова" или "пустой хвост"

это когда head и tail ссылаются на объекты Node, имеющие task = null, то есть не содержащие полезной нагрузки. Поэтому и название "пустая". Это очень удобно, когда head и tail никогда не null. Добавление и удаление новых нод сводится к простому присваиванию ссылок, без необходимости проверять на null, пустоту списка. Обход списка будет начинаться с head.next

Yang-Pi reviewed (commented):
хорошая работа, оставил несколько замечаний и рекомендаций

IvanGriko committed:
ФЗ 6.1
- В Node указать в качестве параметра неопределенный тип T;
- Исправить методов remove() и removeNode() в InMemoryHistoryManager;
- Добавить автотесты

IvanGriko committed:
Create api-tests.yml

IvanGriko committed:
Update InMemoryHistoryManagerTest.java
